### PR TITLE
[mqtt] MQTT binding failed to connect self-signed broker.

### PR DESCRIPTION
**Pull request by falkena**
Fix failed connection to mqtt broker with self-signed certificates.
Fix some java compiler warnings.

Signed-off-by: Alexander Falkenstern <alexander.falkenstern@gmail.com>

### CODE REVIEW

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java [hunk 1]
```diff
@@ -179,40 +178,40 @@ protected void assignSSLContextProvider(BrokerHandlerConfig config, MqttBrokerCo
...
-                config.clientID);
+        MqttBrokerConnection connection = new MqttBrokerConnection(host, config.port, config.secure, config.clientID);
+        connection.setQos(config.qos);
+        connection.setRetain(config.retainMessages);
 
-        final String username = config.username;
-        final String password = config.password;
-        if (StringUtils.isNotBlank(username) && password != null) {
-            connection.setCredentials(username, password); // Empty passwords are allowed
+        if (StringUtils.isNotBlank(config.username) && (config.password != null)) {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java]: You can't use config properties directly. Otherwise our static null checker will warn. Makes sense actually. In a multi threaded environment another thread could alter `config.username` between the check and the usage.

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java [hunk 2]
```diff
@@ -179,40 +178,40 @@ protected void assignSSLContextProvider(BrokerHandlerConfig config, MqttBrokerCo
...
+            connection.setCredentials(config.username, config.password); // Empty passwords are allowed
         }
 
-        final String topic = config.lwtTopic;
+        String topic = config.lwtTopic;
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java]: You are not assigning anything to `topic` are you? So why removing the `final`?

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java [hunk 3]
```diff
@@ -179,40 +178,40 @@ protected void assignSSLContextProvider(BrokerHandlerConfig config, MqttBrokerCo
...
     protected MqttBrokerConnection createBrokerConnection() throws IllegalArgumentException {
         String host = config.host;
-        if (StringUtils.isBlank(host) || host == null) {
+        if ((host == null) || StringUtils.isBlank(host)) {
             throw new IllegalArgumentException("Host is empty!");
         }
 
-        final MqttBrokerConnection connection = new MqttBrokerConnection(host, config.port, config.secure,
-                config.clientID);
+        MqttBrokerConnection connection = new MqttBrokerConnection(host, config.port, config.secure, config.clientID);
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java]: Why did you remove the `final`? It helps other developers to not accidentally overwrite the object

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java [hunk 4]
```diff
@@ -22,20 +21,19 @@
...
 /**
- * This is a custom {@link X509ExtendedTrustManager}. {@link Pin} objects can be added and will
+ * This is a custom {@link X509TrustManager}. {@link Pin} objects can be added and will
  * be used in the checkServerTrusted() method to determine if a connection can be trusted.
  *
  * @author David Graeff - Initial contribution
  */
 @NonNullByDefault
-public class PinTrustManager extends X509ExtendedTrustManager {
+public class PinTrustManager implements X509TrustManager {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: Why did you downgrade to a `X509TrustManager`?

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinningSSLContextProvider.java [hunk 5]
```diff
@@ -42,7 +42,7 @@ public PinningSSLContextProvider(PinTrustManager pinTrustManager) {
     @Override
     public SSLContext getContext() throws ConfigurationException {
         try {
-            SSLContext sslContext = SSLContext.getInstance("TLSv1.2");
+            SSLContext sslContext = SSLContext.getInstance("TLSv1");
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinningSSLContextProvider.java]: Version 1.2 was there on purpose. Everything below is not secure anymore. The version is only a minimum, so 1.3 would be fine as well.

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java [hunk 6]
```diff
@@ -133,32 +112,29 @@ public void checkServerTrusted(X509Certificate @Nullable [] chainN, @Nullable St
...
+    /**
+     * A signature name depends on the security provider but usually follows
+     * https://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#Signature.
+     * E.g.: "SHA256withRSA". We need "SHA" and "256" to initialize a {@link PinMessageDigest}.
+     */
+    PinMessageDigest getMessageDigestForSigAlg(String sigAlg) throws CertificateException {
+        final Matcher matcher = Pattern.compile("(\\D*)(\\d+)").matcher(sigAlg);
+        matcher.find();
+        try {
+            return new PinMessageDigest(matcher.group(1) + "-" + matcher.group(2));
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: This is two lines shorter but nowbody knowns what group1 and group2 stands for like in the previous version.

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java [hunk 7]
```diff
@@ -54,52 +52,33 @@ public void setCallback(PinnedCallback callback) {
...
     }
 
     @Override
-    public void checkServerTrusted(X509Certificate @Nullable [] chainN, @Nullable String authType)
+    public void checkServerTrusted(X509Certificate @Nullable [] certs, @Nullable String authType)
             throws CertificateException {
-        X509Certificate[] chain = chainN;
-        if (chain == null) {
-            return;
+        if ((certs == null) || (certs.length == 0)) {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: There was a reason for this variable. It must be null checked so that our static null checker knows that it is safe to be used in the methods below. In the signature you'll see that `certs` can be null.

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java [hunk 8]
```diff
@@ -54,52 +52,33 @@ public void setCallback(PinnedCallback callback) {
...
-    protected byte[] getEncoded(PinType type, X509Certificate cert) throws CertificateEncodingException {
-        switch (type) {
-            case CERTIFICATE_TYPE:
-                return cert.getEncoded();
-            case PUBLIC_KEY_TYPE:
-                return cert.getPublicKey().getEncoded();
+    @Override
+    public void checkClientTrusted(X509Certificate @Nullable [] certs, @Nullable String authType)
+            throws CertificateException {
+        if ((certs == null) || (certs.length == 0)) {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: Hm, is it really necessary to differentiate between `IllegalArgumentException` and `UnsupportedOperationException`?

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java (hunk 2 above)
Comment (falkena) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java]: Well, 
grafik
Would you guys agree about "final"?

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java (hunk 3 above)
Comment (falkena) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java]: see above

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java (hunk 4 above)
Comment (falkena) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: Since interface of X509TrustManager was used only

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java (hunk 2 above)
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/handler/BrokerHandler.java]: Yup, I know we maintainers have different opinions regarding this detail. That's why I approve new bindings that do not follow this style. But here you are changing existing code :)

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java (hunk 8 above)
Comment (falkena) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: grafik

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java (hunk 7 above)
Comment (falkena) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: As far i can see, certs is still check against null. Miss i something?
grafik

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java (hunk 8 above)
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: Yes If we would implement this method you are right. But we are throwing `UnsupportedOperationException`. If we'd stick to the API contract we'd need to throw `CertificateException` I guess. 

Your changes are not wrong, I'm not arguing with you. But for better keeping track of project changes it helps if you leave unrelated code parts for a particular problem unchanged. And in this case the operation is really just unsupported, no matter if certificates are presented or not. Just saying, leave this in now if you want to.

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java (hunk 4 above)
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: Yes that's reasonable. I haven't found the article I was looking for but for now look at this reference: https://log.rowanto.com/java-8-turning-off-ssl-certificate-check/

The JRE behaves differently depending on if you present a `X509ExtendedTrustManager` or a `X509TrustManager`. And we want the full control here and avoid framework guess work.

File: addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java (hunk 7 above)
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: Just judging by logic, it should work in this short variant. There was a time when the static null checker required a declared non @Nullable variable. Have you also build with maven (/checked the travis output)? The output contains the static code checker results. I'd say we risk it and take this new code.

Comment (falkena) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: As far i've seen, there was no finding about this. I'll cross check again.

Comment (falkena) [reply addons/binding/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/ssl/PinTrustManager.java]: Jenkins is happy: https://ci.openhab.org/job/PR-openHAB2-Addons/12398/Static_20Code_20Analysis_20Report/

**Comment by falkena** [comment]
@davidgraeff : I decline this PR, since it works after cache cleanup. Sorry about your stolen time.

**Comment by davidgraeff** [comment]
All good, I like that you went for the fixing road and maybe you spot other parts that can be improved for your and other setups in the future.
